File: src/codes/god-object/ts/rfc_composite.ts
// Component（共通インターフェース）
interface OrganizationUnit {
  getName(): string;
  print(indent?: string): void;
}

// Leaf（個々の要素）
class Employee implements OrganizationUnit {
  constructor(private name: string) {}

  getName(): string {
    return this.name;
  }

  print(indent: string = ""): void {
    console.log(`${indent}- ${this.name}`);
  }
}

// Composite（集合）
class Department implements OrganizationUnit {
  private members: OrganizationUnit[] = [];

  constructor(private name: string) {}

  add(unit: OrganizationUnit) {
    this.members.push(unit);
  }

  getName(): string {
    return this.name;
  }

  print(indent: string = ""): void {
    console.log(`${indent}+ ${this.name}`);
    this.members.forEach((member) => member.print(indent + "  "));
  }
}

// 利用例
const devTeam = new Department("開発部");
devTeam.add(new Employee("Hiroshi"));
devTeam.add(new Employee("Satoshi"));

const qaTeam = new Department("QA部");
qaTeam.add(new Employee("Yuki"));

const company = new Department("MyCompany");
company.add(devTeam);
company.add(qaTeam);

company.print();

File: src/codes/god-object/py/rfc_composite.py
from abc import ABC, abstractmethod

# Component
class OrganizationUnit(ABC):
    @abstractmethod
    def get_name(self) -> str:
        pass

    @abstractmethod
    def print(self, indent: str = ""):
        pass

# Leaf
class Employee(OrganizationUnit):
    def __init__(self, name: str):
        self.name = name

    def get_name(self) -> str:
        return self.name

    def print(self, indent: str = ""):
        print(f"{indent}- {self.name}")

# Composite
class Department(OrganizationUnit):
    def __init__(self, name: str):
        self.name = name
        self.members: list[OrganizationUnit] = []

    def add(self, unit: OrganizationUnit):
        self.members.append(unit)

    def get_name(self) -> str:
        return self.name

    def print(self, indent: str = ""):
        print(f"{indent}+ {self.name}")
        for member in self.members:
            member.print(indent + "  ")

# 利用例
dev_team = Department("開発部")
dev_team.add(Employee("Hiroshi"))
dev_team.add(Employee("Satoshi"))

qa_team = Department("QA部")
qa_team.add(Employee("Yuki"))

company = Department("MyCompany")
company.add(dev_team)
company.add(qa_team)

company.print()
